File: Sources/AgentTitles.swift
import SwiftUI

enum AgentTitle: String, Codable {
    case presenter

    var displayName: String { "Presenter" }
}

enum AgentTitleScopeKind: String, Codable {
    case room, session, stage
}

struct AgentTitleScope: Codable, Equatable {
    let kind: AgentTitleScopeKind
    let reference: String

    enum CodingKeys: String, CodingKey {
        case kind
        case reference = "ref"
    }
}

enum AgentTitlePermission: String, Codable {
    case stagePresent = "stage.present"
}

/// A reference-only capability envelope. Skill/resource contents remain on
/// the signed host and are never embedded in the grant or transferred to UI.
struct AgentTitleGrant: Identifiable, Codable, Equatable {
    let id: String
    let agentId: String
    let title: AgentTitle
    let scope: AgentTitleScope
    let skillBundleRefs: [String]
    let resourceGrantRefs: [String]
    let delegatedAgentIds: [String]
    let permissions: [AgentTitlePermission]
    let grantedAt: Date
    let expiresAt: Date
    var revokedAt: Date?

    func isActive(at date: Date = Date()) -> Bool {
        grantedAt <= date && date < expiresAt && (revokedAt == nil || date < revokedAt!)
    }
}

enum AgentTitleEventKind: String, Codable {
    case granted, transferred, revoked
}

struct AgentTitleEventReceipt: Identifiable, Codable, Equatable {
    let id: String
    let kind: AgentTitleEventKind
    let at: Date
    let fromAgentId: String?
    let toAgentId: String?
    let reason: String?
}

struct DirectorPolicyDescriptor: Equatable {
    let version: String
    let signatureStatus: String
    let exportable: Bool

    static let builtIn = DirectorPolicyDescriptor(
        version: "director-host-v1",
        signatureStatus: "Verified",
        exportable: false)
}

extension AppStore {
    var activePresenterGrant: AgentTitleGrant? {
        titleGrantsByID.values
            .filter { $0.title == .presenter && $0.isActive() }
            .sorted { $0.grantedAt > $1.grantedAt }
            .first
    }

    var activePresenterAgent: Agent? {
        guard let id = activePresenterGrant?.agentId else { return nil }
        return agents.first { $0.id == id }
    }

    var presenterEligibleAgents: [Agent] {
        agents.filter { activeCallPresenterCandidateIDs.contains($0.id) }
    }

    @discardableResult
    func applyTitleGrant(_ grant: AgentTitleGrant, eventId: String? = nil) -> Bool {
        if let current = activePresenterGrant, current.id != grant.id {
            titleMutationError = "Presenter is already assigned to \(displayName(forAgentID: current.agentId)). Transfer it instead."
            return false
        }
        titleGrantsByID[grant.id] = grant
        appendTitleReceipt(AgentTitleEventReceipt(
            id: eventId ?? "local-title-grant-\(grant.id)",
            kind: .granted,
            at: grant.grantedAt,
            fromAgentId: nil,
            toAgentId: grant.agentId,
            reason: nil))
        titleMutationError = nil
        return true
    }

    func applyTitleTransfer(
        fromGrantId: String,
        toGrant: AgentTitleGrant,
        at: Date,
        eventId: String? = nil
    ) {
        let fromAgentId = titleGrantsByID[fromGrantId]?.agentId
        if var old = titleGrantsByID[fromGrantId] {
            old.revokedAt = at
            titleGrantsByID[fromGrantId] = old
        }
        titleGrantsByID[toGrant.id] = toGrant
        appendTitleReceipt(AgentTitleEventReceipt(
            id: eventId ?? "local-title-transfer-\(fromGrantId)-\(toGrant.id)",
            kind: .transferred,
            at: at,
            fromAgentId: fromAgentId,
            toAgentId: toGrant.agentId,
            reason: nil))
        titleMutationError = nil
    }

    func applyTitleRevocation(
        grantId: String,
        at: Date,
        reason: String,
        eventId: String? = nil
    ) {
        guard var grant = titleGrantsByID[grantId] else { return }
        grant.revokedAt = at
        titleGrantsByID[grantId] = grant
        appendTitleReceipt(AgentTitleEventReceipt(
            id: eventId ?? "local-title-revoke-\(grantId)-\(Int(at.timeIntervalSince1970))",
            kind: .revoked,
            at: at,
            fromAgentId: grant.agentId,
            toAgentId: nil,
            reason: reason))
        titleMutationError = nil
    }

    /// Fold `control.leave`: revoke still-live grants for the leaver.
    /// Stage cards are left in place — same as the kernel.
    func applyControlLeave(
        participantId: String,
        at: Date,
        reason: String = "left",
        eventId: String? = nil
    ) {
        let liveIds = titleGrantsByID.values
            .filter { $0.agentId == participantId && $0.isActive(at: at) }
            .map(\.id)
        for grantId in liveIds {
            applyTitleRevocation(
                grantId: grantId,
                at: at,
                reason: reason,
                eventId: eventId.map { "\($0)-leave-\(grantId)" })
        }
    }

    /// Fold `control.end`: revoke every still-live grant. Cards survive.
    func applyControlEnd(
        at: Date,
        reason: String = "ended",
        eventId: String? = nil
    ) {
        let liveIds = titleGrantsByID.values
            .filter { $0.isActive(at: at) }
            .map(\.id)
        for grantId in liveIds {
            applyTitleRevocation(
                grantId: grantId,
                at: at,
                reason: reason,
                eventId: eventId.map { "\($0)-end-\(grantId)" })
        }
    }

    func requestPresenter(to agentId: String, duration: TimeInterval = 15 * 60) {
        guard activeCallPresenterCandidateIDs.contains(agentId),
              agents.contains(where: { $0.id == agentId }) else {
            titleMutationError = "That agent is not eligible to present in this call."
            return
        }
        let now = Date()
        let newGrant = makePresenterGrant(agentId: agentId, at: now, duration: duration)
        if wireStatus.isLive {
            Task { [weak self] in
                guard let self else { return }
                do {
                    if let current = self.activePresenterGrant {
                        try await self.publishPresenterTransfer(from: current, to: newGrant)
                    } else {
                        try await self.publishPresenterGrant(newGrant)
                    }
                } catch {
                    self.titleMutationError = "The host did not accept the Presenter change."
                }
            }
        } else if let current = activePresenterGrant {
            if current.agentId == agentId { return }
            applyTitleTransfer(fromGrantId: current.id, toGrant: newGrant, at: now)
        } else {
            applyTitleGrant(newGrant)
        }
    }

    func revokePresenter(reason: String = "revoked") {
        guard let current = activePresenterGrant else { return }
        if wireStatus.isLive {
            Task { [weak self] in
                do { try await self?.publishPresenterRevocation(current, reason: reason) }
                catch { self?.titleMutationError = "The host did not accept the Presenter revocation." }
            }
        } else {
            applyTitleRevocation(grantId: current.id, at: Date(), reason: reason)
        }
    }

    func activateDefaultPresenterIfNeeded() {
        guard activePresenterGrant == nil,
              let candidate = activeCallPresenterCandidateIDs.first else { return }
        requestPresenter(to: candidate)
    }

    func makePresenterGrant(
        agentId: String,
        at: Date = Date(),
        duration: TimeInterval = 15 * 60
    ) -> AgentTitleGrant {
        AgentTitleGrant(
            id: "grant-presenter-\(agentId)-\(Int(at.timeIntervalSince1970 * 1000))",
            agentId: agentId,
            title: .presenter,
            scope: AgentTitleScope(kind: .stage, reference: wireActiveProgramId ?? "local-stage"),
            skillBundleRefs: ["bundle-presenter-v1"],
            resourceGrantRefs: ["typed-stage"],
            delegatedAgentIds: [],
            permissions: [.stagePresent],
            grantedAt: at,
            expiresAt: at.addingTimeInterval(duration),
            revokedAt: nil)
    }

    func displayName(forAgentID id: String) -> String {
        agents.first { $0.id == id }?.name ?? id
    }

    private func appendTitleReceipt(_ receipt: AgentTitleEventReceipt) {
        guard !titleEventLog.contains(where: { $0.id == receipt.id }) else { return }
        titleEventLog.append(receipt)
        titleEventLog.sort { $0.at < $1.at }
        if titleEventLog.count > 200 { titleEventLog.removeFirst(titleEventLog.count - 200) }
    }
}

struct PresenterTitleControl: View {
    @EnvironmentObject private var store: AppStore
    @State private var showingDetails = false
    var compact = false

    var body: some View {
        Button { showingDetails = true } label: {
            HStack(spacing: 6) {
                Image(systemName: "rectangle.inset.filled.and.person.filled")
                    .font(.system(size: compact ? 9 : 10, weight: .semibold))
                if let agent = store.activePresenterAgent {
                    Text("Presenter · \(agent.name)")
                } else {
                    Text("Assign Presenter")
                }
            }
            .font(.system(size: compact ? 10.5 : 11.5, weight: .bold))
            .foregroundStyle(.white.opacity(0.88))
            .padding(.horizontal, compact ? 9 : 11)
            .frame(height: compact ? 28 : 32)
            .background(DT.raised, in: Capsule())
            .overlay(Capsule().strokeBorder(.white.opacity(0.12), lineWidth: 0.8))
        }
        .buttonStyle(Pressable())
        .accessibilityHint("Opens the temporary title grant, transfer, and audit controls")
        .sheet(isPresented: $showingDetails) { PresenterControlSheet() }
    }
}

struct PresenterControlSheet: View {
    @EnvironmentObject private var store: AppStore
    @Environment(\.dismiss) private var dismiss
    @Environment(\.colorScheme) private var scheme

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 18) {
                    currentGrant
                    Eyebrow("ELIGIBLE AGENTS")
                    VStack(spacing: 0) {
                        ForEach(store.presenterEligibleAgents) { agent in
                            Button { store.requestPresenter(to: agent.id) } label: {
                                HStack(spacing: 11) {
                                    AvatarChip(letter: String(agent.name.prefix(1)), color: agent.color, size: 30)
                                    VStack(alignment: .leading, spacing: 2) {
                                        Text(agent.name).font(.system(size: 13.5, weight: .bold))
                                        Text("Temporary typed-stage permission")
                                            .font(.system(size: 10.5)).foregroundStyle(.secondary)
                                    }
                                    Spacer()
                                    Image(systemName: store.activePresenterGrant?.agentId == agent.id
                                          ? "checkmark.circle.fill" : "arrow.left.arrow.right.circle")
                                        .foregroundStyle(DT.violet)
                                }
                                .foregroundStyle(DT.ink(scheme))
                                .padding(12)
                                .contentShape(Rectangle())
                            }
                            .buttonStyle(.plain)
                        }
                        if store.presenterEligibleAgents.isEmpty {
                            Text("No Presenter candidates were allowed by this call preset.")
                                .font(.system(size: 12)).foregroundStyle(.secondary).padding(14)
                        }
                    }
                    .card()

                    directorBoundary

                    if !store.titleEventLog.isEmpty {
                        Eyebrow("TITLE EVENTS")
                        VStack(spacing: 0) {
                            ForEach(store.titleEventLog.suffix(8).reversed()) { receipt in
                                HStack(spacing: 9) {
                                    Image(systemName: receipt.kind == .granted ? "plus.circle" : receipt.kind == .transferred ? "arrow.left.arrow.right.circle" : "xmark.circle")
                                        .foregroundStyle(DT.violetText(scheme))
                                    Text(eventLabel(receipt))
                                        .font(.system(size: 11.5, weight: .semibold))
                                    Spacer()
                                    Text(receipt.at.formatted(date: .omitted, time: .shortened))
                                        .font(.system(size: 10, design: .monospaced)).foregroundStyle(.secondary)
                                }
                                .padding(11)
                            }
                        }
                        .card()
                    }

                    if let error = store.titleMutationError {
                        Label(error, systemImage: "exclamationmark.triangle")
                            .font(.system(size: 11.5, weight: .semibold))
                            .foregroundStyle(DT.danger)
                    }
                }
                .padding(20)
            }
            .background(DT.page(scheme).ignoresSafeArea())
            .navigationTitle("Presenter")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .confirmationAction) { Button("Done") { dismiss() } }
            }
        }
        .presentationDetents([.large])
    }

    private var currentGrant: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack {
                Label("Presenter", systemImage: "rectangle.inset.filled.and.person.filled")
                    .font(.system(size: 15, weight: .heavy))
                Spacer()
                Text(store.activePresenterAgent?.name ?? "Unassigned")
                    .font(.system(size: 13, weight: .bold)).foregroundStyle(DT.violetText(scheme))
            }
            if let grant = store.activePresenterGrant {
                TimelineView(.periodic(from: .now, by: 1)) { context in
                    let seconds = max(0, Int(grant.expiresAt.timeIntervalSince(context.date)))
                    Text("Expires in \(seconds / 60)m \(seconds % 60)s · \(grant.scope.kind.rawValue) scope")
                        .font(.system(size: 11, design: .monospaced)).foregroundStyle(.secondary)
                }
                Text("\(grant.skillBundleRefs.count) skill bundle ref · \(grant.resourceGrantRefs.count) resource ref · stage.present")
                    .font(.system(size: 10.5)).foregroundStyle(.secondary)
                Button(role: .destructive) { store.revokePresenter() } label: {
                    Text("Revoke Presenter").font(.system(size: 12, weight: .bold))
                }
            } else {
                Text("Only one agent can own the typed stage. Assigning creates a temporary, replayable event.")
                    .font(.system(size: 11.5)).foregroundStyle(.secondary)
            }
            Label("Typed work only — no device screen, camera, or pixel stream", systemImage: "rectangle.on.rectangle.slash")
                .font(.system(size: 10.5, weight: .semibold)).foregroundStyle(.secondary)
        }
        .padding(14).card()
    }

    private var directorBoundary: some View {
        let policy = DirectorPolicyDescriptor.builtIn
        return VStack(alignment: .leading, spacing: 7) {
            Label("Director policy boundary", systemImage: "checkmark.shield")
                .font(.system(size: 13.5, weight: .bold))
            Text("\(policy.version) · \(policy.signatureStatus) · Host-only")
                .font(.system(size: 10.5, design: .monospaced)).foregroundStyle(.secondary)
            Text("The title passes opaque bundle and resource references. Prompts, routing, tools, scoring, and model configuration stay signed and non-exportable on the host.")
                .font(.system(size: 11.5)).foregroundStyle(.secondary)
        }
        .padding(14).card()
    }

    private func eventLabel(_ receipt: AgentTitleEventReceipt) -> String {
        switch receipt.kind {
        case .granted:
            return "Granted to \(agentName(receipt.toAgentId))"
        case .transferred:
            return "\(agentName(receipt.fromAgentId)) → \(agentName(receipt.toAgentId))"
        case .revoked:
            return "Revoked from \(agentName(receipt.fromAgentId))"
        }
    }

    private func agentName(_ id: String?) -> String {
        guard let id else { return "Agent" }
        return store.displayName(forAgentID: id)
    }
}

File: web/src/views/work/AgentTitles.js
// Presenter title views — a port of the view half of `Sources/AgentTitles.swift`
// (`PresenterTitleControl`, `PresenterControlSheet`). The store half (grants,
// transfer, revocation, receipts) lives in store.js.
//
// A grant is a reference-only capability envelope: this UI shows the agent,
// the scope, the expiry and the counts of opaque refs — never skill or prompt
// contents, which stay signed and non-exportable on the host.
import { html, cx, useTick, useObservable } from "../../ui.js";
import { Icon, AvatarChip, NavBar, Eyebrow } from "../../components.js";
import { nav } from "../../nav.js";
import { DIRECTOR_POLICY } from "../../models.js";
import { injectStyle, CheckCircle } from "./shared.js";

let store = null;
export function bindAgentTitles(ctx) { store = ctx.store; }

injectStyle("work-css-titles", `
.pt-chip { display: inline-flex; align-items: center; gap: 6px; height: 32px; padding: 0 11px; border-radius: 999px; font-size: 11.5px; font-weight: 700;
  color: rgba(var(--ink-rgb), .88); background: var(--surface); box-shadow: inset 0 0 0 0.8px var(--hairline); white-space: nowrap; }
.pt-chip.compact { height: 28px; padding: 0 9px; font-size: 10.5px; }
.pt-chip.unclaimed { color: var(--violet-text); }
.pt-card { padding: 14px; display: flex; flex-direction: column; gap: 10px; }
.pt-card .head { display: flex; align-items: center; gap: 8px; font-size: 15px; font-weight: 800; }
.pt-card .holder { margin-left: auto; font-size: 13px; font-weight: 700; color: var(--violet-text); }
.pt-card .meta { font-family: var(--mono); font-size: 11px; color: var(--ink-55); }
.pt-card .refs { font-size: 10.5px; color: var(--ink-55); }
.pt-card .body { font-size: 11.5px; color: var(--ink-55); line-height: 1.4; }
.pt-card .boundary { display: flex; align-items: center; gap: 7px; font-size: 10.5px; font-weight: 600; color: var(--ink-55); }
.pt-revoke { align-self: flex-start; min-height: 44px; display: inline-flex; align-items: center; padding: 0 12px; margin: -6px 0 -6px -12px; border-radius: var(--r-control); font-size: 12px; font-weight: 700; color: var(--danger); }
.pt-agent { display: flex; align-items: center; gap: 11px; width: 100%; padding: 12px; text-align: left; color: var(--ink); min-height: 54px; }
.pt-agent + .pt-agent { box-shadow: inset 0 0.8px 0 var(--hairline); }
.pt-agent:active { background: var(--surface2); }
.pt-agent .n { font-size: 13.5px; font-weight: 700; }
.pt-agent .s { font-size: 10.5px; color: var(--ink-55); margin-top: 2px; }
.pt-agent .ic { margin-left: auto; color: var(--violet); display: grid; flex: none; }
.pt-none { font-size: 12px; color: var(--ink-55); padding: 14px; line-height: 1.4; }
.pt-event { display: flex; align-items: center; gap: 9px; padding: 11px; font-size: 11.5px; font-weight: 600; min-height: 44px; }
.pt-event + .pt-event { box-shadow: inset 0 0.8px 0 var(--hairline); }
.pt-event .ic { color: var(--violet-text); display: grid; flex: none; }
.pt-event .when { margin-left: auto; font-family: var(--mono); font-size: 10px; color: var(--ink-55); flex: none; }
.pt-event .why { color: var(--ink-35); font-weight: 500; }
.pt-director { padding: 14px; display: flex; flex-direction: column; gap: 7px; }
.pt-director .head { display: flex; align-items: center; gap: 7px; font-size: 13.5px; font-weight: 700; }
.pt-director .meta { font-family: var(--mono); font-size: 10.5px; color: var(--ink-55); }
.pt-director .body { font-size: 11.5px; color: var(--ink-55); line-height: 1.4; }
`);

const timeFmt = new Intl.DateTimeFormat([], { hour: "numeric", minute: "2-digit" });

/** The chip in the live call: who holds Presenter right now (or "Assign Presenter"). */
export function PresenterTitleControl({ compact = false, className }) {
  const s = useObservable(store ?? window.drive.store);
  const agent = s.activePresenterAgent;
  return html`<button type="button" class=${cx("pt-chip pressable", compact && "compact", !agent && "unclaimed", className)}
    onClick=${() => nav.present("presenterControl")}
    aria-label=${agent ? `Presenter, ${agent.name}` : "Assign Presenter"} title="Opens the temporary title grant, transfer, and audit controls">
    <${Icon} name="rectangle.inset.filled.and.person.filled" size=${compact ? 11 : 12} weight=${2.4} />
    <span>${agent ? `Presenter · ${agent.name}` : "Assign Presenter"}</span>
  </button>`;
}

function GrantExpiry({ grant }) {
  useTick(1);
  const seconds = Math.max(0, Math.floor((grant.expiresAt - Date.now()) / 1000));
  return html`<div class="meta">Expires in ${Math.floor(seconds / 60)}m ${seconds % 60}s · ${grant.scope.kind} scope</div>`;
}

function eventLabel(s, r) {
  const name = (id) => (id ? s.displayNameForAgent(id) : "Agent");
  switch (r.kind) {
    case "granted": return `Granted to ${name(r.toAgentId)}`;
    case "transferred": return `${name(r.fromAgentId)} → ${name(r.toAgentId)}`;
    case "revoked": return `Revoked from ${name(r.fromAgentId)}`;
    default: return r.kind;
  }
}
const EVENT_ICON = { granted: "plus.circle", transferred: "arrow.left.arrow.right.circle", revoked: "xmark.circle" };

/** Route "presenterControl" — the grant, transfer, revoke and audit sheet. */
export function PresenterControlSheet() {
  const s = useObservable(store ?? window.drive.store);
  const grant = s.activePresenterGrant;
  const holder = s.activePresenterAgent;
  const eligible = s.presenterEligibleAgents;
  const events = s.titleEventLog.slice(-8).reverse();
  return html`<div class="wk-sheet">
    <${NavBar} title="Presenter" trailing=${html`<button class="btn ghost sm" onClick=${() => nav.dismiss()}>Done</button>`} />
    <div class="scroll"><div class="content vstack" style=${{ "--gap": "18px", paddingTop: 4 }}>
      <div class="card pt-card">
        <div class="head"><${Icon} name="rectangle.inset.filled.and.person.filled" size=${16} weight=${2.3} /> Presenter
          <span class="holder">${holder?.name ?? "Unassigned"}</span></div>
        ${grant ? html`
          <${GrantExpiry} grant=${grant} />
          <div class="refs">${grant.skillBundleRefs.length} skill bundle ref · ${grant.resourceGrantRefs.length} resource ref · ${grant.permissions.join(", ")}</div>
          <button type="button" class="pt-revoke pressable" onClick=${() => s.revokePresenter()}>Revoke Presenter</button>
        ` : html`<div class="body">Only one agent can own the typed stage. Assigning creates a temporary, replayable event.</div>`}
        <div class="boundary"><${Icon} name="rectangle.on.rectangle.slash" size=${13} weight=${2.2} /> Typed work only — no device screen, camera, or pixel stream</div>
      </div>

      <div>
        <${Eyebrow} style=${{ padding: "0 4px 8px" }}>ELIGIBLE AGENTS</${Eyebrow}>
        <div class="card" style=${{ overflow: "hidden" }}>
          ${eligible.map((agent) => {
            const holds = grant?.agentId === agent.id;
            return html`<button key=${agent.id} type="button" class="pt-agent" onClick=${() => s.requestPresenter(agent.id)}
              aria-label=${holds ? `${agent.name}, current Presenter` : `${grant ? "Transfer Presenter to" : "Grant Presenter to"} ${agent.name}`}>
              <${AvatarChip} name=${agent.name} color=${agent.color} size=${30} />
              <div><div class="n">${agent.name}</div><div class="s">Temporary typed-stage permission</div></div>
              <span class="ic">${holds ? html`<${CheckCircle} on size=${18} />` : html`<${Icon} name="arrow.left.arrow.right.circle" size=${18} />`}</span>
            </button>`;
          })}
          ${!eligible.length ? html`<div class="pt-none">No Presenter candidates were allowed by this call preset.</div>` : null}
        </div>
      </div>

      <div class="card pt-director">
        <div class="head"><${Icon} name="checkmark.shield" size=${15} weight=${2.3} /> Director policy boundary</div>
        <div class="meta">${DIRECTOR_POLICY.version} · ${DIRECTOR_POLICY.signatureStatus} · ${DIRECTOR_POLICY.exportable ? "Exportable" : "Host-only · not exportable"}</div>
        <div class="body">The title passes opaque bundle and resource references. Prompts, routing, tools, scoring, and model configuration stay signed and non-exportable on the host.</div>
      </div>

      ${events.length ? html`<div>
        <${Eyebrow} style=${{ padding: "0 4px 8px" }}>TITLE EVENTS</${Eyebrow}>
        <div class="card" style=${{ overflow: "hidden" }}>
          ${events.map((r) => html`<div key=${r.id} class="pt-event">
            <span class="ic"><${Icon} name=${EVENT_ICON[r.kind] ?? "circle"} size=${16} /></span>
            <span>${eventLabel(s, r)}${r.reason ? html` <span class="why">· ${r.reason}</span>` : null}</span>
            <span class="when">${timeFmt.format(r.at)}</span>
          </div>`)}
        </div>
      </div>` : null}

      ${s.titleMutationError ? html`<div class="wk-error" role="alert"><${Icon} name="exclamationmark.triangle" size=${14} weight=${2.3} /> ${s.titleMutationError}</div>` : null}
    </div></div>
  </div>`;
}
